lib/class-wp-theme-json-gutenberg.php

```diff
@@ -838,55 +838,21 @@ protected static function get_blocks_metadata() {
 		}
 
 		foreach ( $blocks as $block_name => $block_type ) {
-			if (
-				isset( $block_type->supports['__experimentalSelector'] ) &&
-				is_string( $block_type->supports['__experimentalSelector'] )
-			) {
-				static::$blocks_metadata[ $block_name ]['selector'] = $block_type->supports['__experimentalSelector'];
-			} else {
-				static::$blocks_metadata[ $block_name ]['selector'] = '.wp-block-' . str_replace( '/', '-', str_replace( 'core/', '', $block_name ) );
-			}
+			$root_selector = wp_get_block_css_selector( $block_type );
 
-			if (
-				isset( $block_type->supports['color']['__experimentalDuotone'] ) &&
-				is_string( $block_type->supports['color']['__experimentalDuotone'] )
-			) {
-				static::$blocks_metadata[ $block_name ]['duotone'] = $block_type->supports['color']['__experimentalDuotone'];
-			}
-
-			// Generate block support feature level selectors if opted into
-			// for the current block.
-			$features = array();
-			foreach ( static::BLOCK_SUPPORT_FEATURE_LEVEL_SELECTORS as $key => $feature ) {
-				if (
-					isset( $block_type->supports[ $key ]['__experimentalSelector'] ) &&
-					$block_type->supports[ $key ]['__experimentalSelector']
-				) {
-					$features[ $feature ] = static::scope_selector(
-						static::$blocks_metadata[ $block_name ]['selector'],
-						$block_type->supports[ $key ]['__experimentalSelector']
-					);
-				}
-			}
+			static::$blocks_metadata[ $block_name ]['selector']  = $root_selector;
+			static::$blocks_metadata[ $block_name ]['selectors'] = static::get_block_selectors( $block_type, $root_selector );
 
-			if ( ! empty( $features ) ) {
-				static::$blocks_metadata[ $block_name ]['features'] = $features;
+			$elements = static::get_block_element_selectors( $root_selector );
+			if ( ! empty( $elements ) ) {
+				static::$blocks_metadata[ $block_name ]['elements'] = $elements;
 			}
 
-			// Assign defaults, then overwrite those that the block sets by itself.
-			// If the block selector is compounded, will append the element to each
-			// individual block selector.
-			$block_selectors = explode( ',', static::$blocks_metadata[ $block_name ]['selector'] );
-			foreach ( static::ELEMENTS as $el_name => $el_selector ) {
-				$element_selector = array();
-				foreach ( $block_selectors as $selector ) {
-					if ( $selector === $el_selector ) {
-						$element_selector = array( $el_selector );
-						break;
-					}
-					$element_selector[] = static::append_to_selector( $el_selector, $selector . ' ', 'left' );
-				}
-				static::$blocks_metadata[ $block_name ]['elements'][ $el_name ] = implode( ',', $element_selector );
+			// The block may or may not have a duotone selector.
+			// TODO: Should this target be `color.duotone` not `duotone`?
```

Comment:
I think the idea is to move `duotone` to its own config outside of `color`.

Reply:
Thanks for the heads up 👍 

@ajlende What are your thoughts on where `duotone` gets located within the `selectors` object? 

Would you prefer it to be located under `filters.duotone` as per the block support proposal in https://github.com/WordPress/gutenberg/pull/42870, or standalone as `duotone`? Given the possibilities for further filters, I'm leaning towards `filters.duotone` myself.

Reply:
I'd say `filters.duotone` for sure 👍

Reply:
@getdave @ajlende @aaronrobertshaw I may have missed tons of conversations, so I'll be happy to update my understanding. I've only learned [now](https://github.com/WordPress/gutenberg/pull/49103#discussion_r1147091806) that we went with `filters.duotone` here. My current understanding is that we tried to use the same object shape in the following places:

- theme.json keys
- block style attribute
- block supports

Given that we use `color.duotone` in those places, why the new selectors field should be any different?

Reply:
Alex and I talked about this privately. This is the current situation:

- theme.json `settings.color.duotone`
- theme.json `styles.filter.duotone`
- block.json supports `color.duotone`
- block style attribute `style.color.duotone`
- block.json selectors `style.filters.duotone`

Follow-ups:

- https://github.com/WordPress/gutenberg/issues/49342 
- https://github.com/WordPress/gutenberg/issues/49340

Reply:
So my current understanding is that we'll need to tweak the Selectors API for duotone to be under `selectors.filter.duotone` using the singular `filter` instead of `filters`. 

That's an easy change from the API's end, I'll make it happen. It will also impact https://github.com/WordPress/gutenberg/pull/49325.